protocol/proofs/withdrawal-invalidation-mitigation.md

```diff
@@ -0,0 +1,83 @@
+defines a variable `respectedGameType` which determines the particular `DisputeGame` code that
+users can utilize when proving and finalizing withdrawals. The "fallback" safety net action is the
+ability for the Guardian or Deputy Guardian address to change the `respectedGameType` variable to
+some alternative game type (e.g., the `PermissionedDisputeGame`) in the case that the
+`FaultDisputeGame` contract is found to be buggy.
+
+## Problem Statement
+
+The fallback action of changing the `respectedGameType` variable updates another variable called
+`respectedGameTypeUpdatedAt` that keeps track of exactly when the respected game type was changed.
+Currently, the `OptimismPortal` contract enforces that withdrawals can only be proven or finalized
+against `DisputeGame` contracts that were created with a timestamp greater than the
+`respectedGameTypeUpdatedAt` variable. This effectively invalidates any withdrawal proofs that were
+submitted prior to the activation of the fallback mechanism.
+
+Withdrawal invalidation is a powerful tool in the incident response playbook as it can mitigate the
+impact of a large number of invalid dispute games at the same time. However, it also forces
+users to resubmit their withdrawal proofs and wait an additional 7 days to execute a withdrawal.
+Users have reported that this invalidation and additional delay period is a pain point. This user
+impact means that the fallback is highly disruptive as a standard security mechanism.
+
+We are in the process of further assessing the impact and severity of this pain point but are
+creating a design document in the meanwhile to determine the lift to resolve this issue.
+
+## Proposed Solution
+
+We propose solving this problem by making withdrawal invalidation an optional part of the fallback
+mechanism. We can achieve this by slightly modifying the rules for withdrawal verification.
+
+We currently enforce the following set of rules (not including other rules like blacklist checks):
+
+- Withdrawals must be proven against a dispute game of the `respectedGameType`.
+- Withdrawals must be finalized against a dispute game of the `respectedGameType`.
+- Withdrawals must be finalized against a dispute game created after `respectedGameTypeUpdatedAt`.
+
+We propose removing the requirement that withdrawals must be finalized against a dispute game of
+the `respectedGameType`. Withdrawals that have been previously proven against a dispute game when
+that game *was* respected can be finalized even if that dispute game is no longer of the respected
+game type.
+
+We additionally propose that the `respectedGameTypeUpdatedAt` timestamp not be updated by default.
+Instead, the timestamp can be selectively updated if withdrawal invalidation is actually required.
+Combined with the first proposed change, this proposal means that older withdrawal proofs can
+usually be finalized when the respected game type is changed but withdrawals can still be blanket
+invalidated if necessary.
```

Comment:
An issue can possibly arise if the second case for the withdrawal invalidation without checking the `respectGameType` like in the past.

For example, this might be a possible attack vector:

1. The guardian make a tx to set the `respectedGameType` to 1 (permissioned game).
2. The attacker monitors the mempool and frontruns **in the same block** the call to `setRespectGame(1)`.
3. In the same block, the attacker create a game using `create()` on the Fault Dispute Factory contract. Because the `createAt` is a timestamp from (`block.timestamp`).
4. The attacker call `proveWithdrawalTransaction()` and then wait for the `PROOF_MATURITY_DELAY_SECONDS` submit the `finalizeWithdrawalTransaction()`.
5. As the check (below in solidity) is using ` createdAt >= respectedGameTypeUpdatedAt` the attacker might be able to finalize a transaction with the incorrect `gameType`.
6. The impact seems to be really low, but this would make an inconsistent state (as the `optimismPortal` and the operator will think that only certain gameType are valid during the withdrawal process).

```solidity

if (disputeGameProxy.gameType().raw() != respectedGameType.raw()) //@audit: The current design review propose to not use this anymore?
    revert InvalidGameType();

require(
    createdAt >= respectedGameTypeUpdatedAt, //@audit: Usage of the >= instead of > is a potential issue here?
    "OptimismPortal: dispute game created before respected game type was updated"
);

```

This scenario might be invalid, don't hesitate to let me know about it.

Reply:
Yeah would be worth making that a strict `>`.  While that will mean that any withdrawals proven against the new game type after the change but in the same block are invalidated when they probably shouldn't be, that's a very unlikely corner case and much lower impact than having a withdrawal be unexpectedly valid.

Reply:
Let’s make sure that we account for this in the spec.

Reply:
> While that will mean that any withdrawals proven against the new game type after the change but in the same block are invalidated when they probably shouldn't be, that's a very unlikely corner case and much lower impact than having a withdrawal be unexpectedly valid.

We can get around this by reverting if someone tries to prove a withdrawal within the same block. Better to have an edge-case revert where you get immediate feedback than an edge-case invalid withdrawal where you only get feedback later.